README.md

Educational Aggregator for CARRE project
=========
The aim of the educational resource aggregator is to harvest educational resources from 3rd party repositories, present these to the medical expert for annotation and rating, and output the results of the annotation (together with resource metadata) to the CARRE public RDF repository.

The main parts of this aggregator are: the Resource Retriever, the Resource Rating, the Resource Metadata Processing, and the User Application.

The Resource Retriever accepts CARRE concept terms from the CARRE public RDF repository and uses them to formulate queries to external 3rd party educational resource repositories. The results of this search are parsed to extract metadata. Then the retrieved results and metadata are displayed to the expert user for rating and annotation (via the aggregator front end). The module consists of 2 services that make use of SPARQL protocol in the case of query term extraction from the CARRE server and API requests to each educational repository.
The Resource Rating module allows the input of expert user opinion and annotation, and also calculates subjective scores that measure the quality of the resource. Expert rating involves assessment of content-keyword relevance, content accuracy and depth of coverage, while the automatic systems rating is based on the Readability Test of the Flesch-Kincaid algorithm, and rating based on the latest modified version of the article and number of revisions. The module is an optional process that requires special user privileges. Each user role is assigned to different rating criteria according to user authorization. As of version 0.2, only 2 user roles have been taken into implementation: (a) the expert, and (b) the general public.
The Resource Metadata Processing module involves metadata enrichment via semantic web sources (such as NCBO BioPortal medical ontologies and DBpedia). The module is a combination of 3 services that collect data per article, making multiple SPARQL requests to enrich the data and finally store it as a unique identified resource into the local MongoDB datastore. Then data is transformed into RDF triples in order to be inserted to CARRE educational repository.
The User Application is a web application accessible at http://edu.carre-project.eu/. The visible components of the web application are built upon html5 and CSS3 using modern frameworks for consistency and responsiveness like Twitter Bootstrap CSS Framework have been extensively used. Bootstrap is also responsible for mobile/tablet view of the web application.




Public URL's
-------

Development url : http://beta.carre-project.eu:8080

Code metrics url : http://beta.carre-project.eu:9999

Production url : http://edu.carre-project.eu

Testing and Deployment services
---------
[![Build Status](https://travis-ci.org/telemed-duth/carre-edu.svg?branch=master)](https://travis-ci.org/telemed-duth/carre-edu)
[![Build Status](https://codeship.com/projects/8f8138f0-06cb-0132-1e81-36288a81fed8/status?branch=master)](https://codeship.com/projects/31184)
[![Code Climate](https://codeclimate.com/github/telemed-duth/carre-edu/badges/gpa.svg)](https://codeclimate.com/github/telemed-duth/carre-edu)
[![Test Coverage](https://codeclimate.com/github/telemed-duth/carre-edu/badges/coverage.svg)](https://codeclimate.com/github/telemed-duth/carre-edu)


Technology stack
----------------
    Client : AngularJS , bootstrap css, HTML5
    
    Server : NodeJS, mongodb with mongoose, socket.io for sockets
    
    Deployment : 2-container(app+db) docker with dokku-alt
    
    Workflow : gruntjs,yeoman fullstack generator, plato
    



Host on your own server!
--------------
Minimum Requirements: 1GB RAM + 3GB HDD
The deployment is supported only on a unix* like machine (linux , Mac) and requires the following libraries to installed on your computer : 

* NodeJS application sever
* MongoDB database server
* Git version control system

Next you should clone the repository at github, install all dependencies and run the build script.

You should take into considaration that the dependencies require ~3gb of space.

    The commands for the above steps are self-explanatory :
            
    $ git clone https://github.com/telemed-duth/carre-edu.git
    $ npm install -g bower grunt-cli
    $ npm install && bower install
    $ grunt serve
            

Now the system is running on the default port 8080 and can be accessed at http://localhost:8080 through a supported web browser. 


    
[MIT License (MIT)](https://raw.githubusercontent.com/telemed-duth/carre-edu/master/LICENSE.txt)
------------
Copyright (c) 2014 Democritus University of Thrace (DUTH) , Greece


CHANGELOG.txt
# CHANGELOG v0.3.1

- * 76ecdf5 - 2015-02-22: CLI: bug fixes for multiple rating triples, rating based on 2-way rdf function 
- * 2c4be4f - 2015-02-22: CLI: fetching user and other rated articles 
- * 76415f2 - 2015-02-22: CLI: insert/modify dynamic rating criteria 
- * 7b1d546 - 2015-02-22: CLI: add rank and refactor main service 
- * e89fda6 - 2015-02-21: remove apache conf 
- * 81a0608 - 2015-02-21: CLI: add modify views per article to rdf 
- * 1e9511a - 2015-02-20: CLI: add rating and enrichment service 
- * abb01d9 - 2015-02-19: remove .log files 
- * 68bc6b3 - 2015-02-18: CLI: fix newrelic log 
- * e410615 - 2015-02-18: CLI: article insert through api + token 
- * 2e478f2 - 2015-02-17: CLI: update process got lifting 
- * 39243cd - 2015-02-16: CLI: fix app name 
- * 72646f0 - 2015-02-16: CLI: update dependencies and add article,enrichment,rdf services 
- * 17e9d87 - 2015-02-13: CLI: add change password for CARRE users 
- * bb557ba - 2015-02-13: CLI: update navbar link to CARRE devices 
- * acbad64 - 2015-02-13: CLI: updated settings url when logged in as CARRE 
- * 4910241 - 2015-02-13: CLI: finalize automated versioning system 
- * 3ac05d1 - 2015-02-13: CLI: fix update no 3 
- * 3b3b367 - 2015-02-13: CLI: update script 
- * 7d7765a - 2015-02-13: CLI: update script fix 
- * a7fb7fb - 2015-02-13: CLI: set defaults for the update script 
- * 9345c27 - 2015-02-13: CLI: playing with versioning 
- * 8c630fb - 2015-02-13: CLI: try to configure automated versioning 
- * be830cf - 2015-02-13: CLI: update description and badges 
- * b8c12c4 - 2015-02-13: CLI: fix codeship badge 
- * 4959eef - 2015-02-13: CLI: replace preinstall--&gt; pretest npm scripts 
- * c0fb9dd - 2015-02-13: CLI: fix codeship build try no.1 
- * 9c33807 - 2015-02-13: CLI: add codeship badge and fix deployment to dokku 
- * d32bd1e - 2015-02-13: CLI: add Continuous deployment!! 
- * 593fec0 - 2015-02-13: CLI: update travis and codeship test routines 
- * 3dd9727 - 2015-02-13: CLI: codeship fix.. 
- * 97d9840 - 2015-02-12: CLI: fix codeship build 
- * 958ba59 - 2015-02-12: CLI: add logout function to carre authentication and fix testing failure 
- * 288cb15 - 2015-02-12: CLI: use default icon when gravatar not exist 
- * 3b62f15 - 2015-02-12: CLI: fix gravatar 
- * 17e7828 - 2015-02-12: CLI: update authentication and gravatar generated pic 
- * 8e233c7 - 2015-02-10: CLI: add newrelic profiler 
- * 5e1606a - 2015-02-10: CLI: finalize authentication with the Open University authentication mechanism 
- * a53305d - 2015-02-09: CLI: carre token authentication addon 
- * bc6967b - 2015-02-05: CLI: ui and wikipedia query optimizations 
- * c386a0f - 2015-02-05: CLI: fix no results endless loop bug 
- * 3903b13 - 2015-02-05: CLI: fix medline plus bug 
- * f9d28b2 - 2015-02-04: CLI: fix wikipedia crash bug 
- * 1448945 - 2015-02-03: CLI: change update scripts , add openid fix 
- * c7936ce - 2015-02-03: add openid auth-not working 
- * f7801ec - 2015-02-01: CLI: remove console.log for undefined 
- * 7044aec - 2015-02-01: CLI: undefined search term bug fix , duplicate risk element 
- * 48bba86 - 2015-02-01: CLI: fix bug created by a double quote and update cli tool not to build script twice 
- * 7742311 - 2015-02-01: CLI: add minimum requirements 
- * 5306abb - 2015-02-01: CLI: update login to enable rating and some guidelines on readme file 
- * 53188db - 2015-02-01: CLI: update cli tool command 
- * ac7d120 - 2015-02-01: CLI: fix no results on wikipedia and redirect to medlineplus 
- * e2ce6a2 - 2015-02-01: CLI: fsdfsdfsdfsd 
- * eb46766 - 2015-01-31: BUILD: fix bug when selecting a term autosearch 
- * 6273437 - 2015-01-31: fix bug in riskelements hardcoded data 
- * 106e590 - 2015-01-31: fix medlineplus no results bug and redirect to wikipedia 
- * d6287cd - 2015-01-31: add autogenerated version to changelog 
- * 339b30d - 2015-01-31: fix change log 
- * 2f25944 - 2015-01-31: add auto changelog generation! 
- * 27f531d - 2015-01-31: add auto changelog generation! 
- * 9d27bee - 2015-01-30:  remove npm-log 
- * e1a3935 - 2015-01-30: add code metrics to grunt workflow 
- * 1cec316 - 2015-01-29: fix medlineplus and wikipedia mobile view 
- * bdae6f3 - 2015-01-28: integrate medlineplus to new environment-currently only list is working 
- * 5f14bae - 2015-01-27: deployment instructions 
- * 780b56b - 2015-01-27: add license file link to readme 
- * 5752f9b - 2015-01-27: add license file 
- * a525c63 - 2015-01-27: add code metrics url 
- * 0ef6e70 - 2015-01-27: add code metrics generator and server! 
- * 4f2c9d8 - 2015-01-25: fix preview transition 
- * ab317d8 - 2015-01-25: add rating module 
- * c22802d - 2015-01-25: add sliding panel to article view 
- * d835d16 - 2015-01-24: fix preview window left-right buttons 
- * 2dc4acb - 2015-01-24: add article preview 
- * 47c53bf - 2015-01-24: update development url 
- * 8eb55c6 - 2015-01-23: use global header and footer 
- * f428bf8 - 2015-01-23: remove sidebar from master 
- * 7e83afc - 2015-01-23: add new dev url 
- * 310fef1 - 2015-01-23: add new sidebar 
- * d483726 - 2015-01-23: delete collection 
- * 24921f3 - 2015-01-23: add rdash 
- * 92fc9e7 - 2015-01-21:  update version to 0.0.2 
- * 850b4b3 - 2015-01-21: update npm dependencies 
- * 025e4c0 - 2015-01-21: fix livereload for c9 
- * 4f59de6 - 2014-12-18: add .c9 to .gitignore 
- * db94669 - 2014-12-18: add carre auth dir 
- * 48e928b - 2014-12-18: add oauth carre authentication 
- * 4e7d9ba - 2014-12-05: change riskelements source 
- * 3382cbf - 2014-11-27: BUILD: add rating to medlineplus 
- * cf455b9 - 2014-11-25: BUILD: change color on global search , make more precise wikipedia search , fix home page when not logged in 
- * e9f2908 - 2014-11-24: BUILD: add carre risk elements fetch from sparql , global search finished, changed home to medlineplus 
- * 93c5bf6 - 2014-11-24: BUILD: fix css close on global search , fix reload on same page when global search works 

Gruntfile.js
// Generated on 2014-11-04 using generator-angular-fullstack 2.0.13
'use strict';

module.exports = function (grunt) {
  var localConfig;
  try {
    localConfig = require('./server/config/local.env');
  } catch(e) {
    localConfig = {};
  }

  // Load grunt tasks automatically, when needed
  require('jit-grunt')(grunt, {
    express: 'grunt-express-server',
    useminPrepare: 'grunt-usemin',
    ngtemplates: 'grunt-angular-templates',
    cdnify: 'grunt-google-cdn',
    protractor: 'grunt-protractor-runner',
    injector: 'grunt-asset-injector',
    buildcontrol: 'grunt-build-control'
  });

  // Time how long tasks take. Can help when optimizing build times
  require('time-grunt')(grunt);

  // Define the configuration for all the tasks
  grunt.initConfig({

    // Project settings
    pkg: grunt.file.readJSON('package.json'),
    yeoman: {
      // configurable paths
      client: require('./bower.json').appPath || 'client',
      dist: 'dist'
    },
    run: {
      metrics: {
        options:{
          wait:false
        },
        exec: './metrics',
      }
    },
    changelog: {
      v1: {
        options: {
          fileHeader:'# CHANGELOG v<%= pkg.version%>',
          after: '2014-10-01',
          logArguments: [
            '--pretty=* %h - %ad: %s',
            '--no-merges',
            '--date=short'
          ],
          dest: 'release-notes/Version-<%= pkg.version %>.txt',
          template: '{{date}}\n\n{{> features}}{{> fixes}}',
          featureRegex: /^(.*)$/gim,
          partials: {
            features: '{{#if features}}{{#each features}}{{> feature}}{{/each}}{{else}}{{> empty}}{{/if}}\n',
            feature: '- {{this}} {{this.date}}\n'
          }
        }
      }
    },
    bump: {
      options: {
        files: ['package.json','bower.json'],
        updateConfigs: ['pkg'],
        commit: false,
        commitMessage: 'Release v%VERSION%',
        commitFiles: ['package.json'],
        createTag: true,
        tagName: 'v%VERSION%',
        tagMessage: 'Version %VERSION%',
        push: true,
        pushTo: 'origin',
        gitDescribeOptions: '--tags --always --abbrev=1 --dirty=-d',
        globalReplace: false
      }
    },
  
    express: {
      options: {
        port: process.env.PORT || 9000
      },
      dev: {
        options: {
          script: 'server/app.js',
          debug: true
        }
      },
      prod: {
        options: {
          script: 'dist/server/app.js'
        }
      }
    },
    open: {
      server: {
        url: 'http://localhost:<%= express.options.port %>'
      }
    },
    watch: {
      injectJS: {
        files: [
          '<%= yeoman.client %>/{app,components}/**/*.js',
          '!<%= yeoman.client %>/{app,components}/**/*.spec.js',
          '!<%= yeoman.client %>/{app,components}/**/*.mock.js',
          '!<%= yeoman.client %>/app/app.js'],
        tasks: ['injector:scripts']
      },
      injectCss: {
        files: [
          '<%= yeoman.client %>/{app,components}/**/*.css'
        ],
        tasks: ['injector:css']
      },
      mochaTest: {
        files: ['server/**/*.spec.js'],
        tasks: ['env:test', 'mochaTest']
      },
      jsTest: {
        files: [
          '<%= yeoman.client %>/{app,components}/**/*.spec.js',
          '<%= yeoman.client %>/{app,components}/**/*.mock.js'
        ],
        tasks: ['newer:jshint:all', 'karma']
      },
      gruntfile: {
        files: ['Gruntfile.js']
      },
      livereload: {
        files: [
          '{.tmp,<%= yeoman.client %>}/{app,components}/**/*.css',
          '{.tmp,<%= yeoman.client %>}/{app,components}/**/*.html',
          '{.tmp,<%= yeoman.client %>}/{app,components}/**/*.js',
          '!{.tmp,<%= yeoman.client %>}{app,components}/**/*.spec.js',
          '!{.tmp,<%= yeoman.client %>}/{app,components}/**/*.mock.js',
          '<%= yeoman.client %>/assets/images/{,*//*}*.{png,jpg,jpeg,gif,webp,svg}'
        ],
        options: {
          
          livereload: true
        }
      },
      express: {
        files: [
          'server/**/*.{js,json}'
        ],
        tasks: ['express:dev', 'wait'],
        options: {
          livereload: true,
          nospawn: false //Without this option specified express won't be reloaded
        }
      }
    },

    // Make sure code styles are up to par and there are no obvious mistakes
    jshint: {
      options: {
        jshintrc: '<%= yeoman.client %>/.jshintrc',
        reporter: require('jshint-stylish')
      },
      server: {
        options: {
          jshintrc: 'server/.jshintrc'
        },
        src: [
          'server/**/*.js',
          '!server/**/*.spec.js'
        ]
      },
      serverTest: {
        options: {
          jshintrc: 'server/.jshintrc-spec'
        },
        src: ['server/**/*.spec.js']
      },
      all: [
        '<%= yeoman.client %>/{app,components}/**/*.js',
        '!<%= yeoman.client %>/{app,components}/**/*.spec.js',
        '!<%= yeoman.client %>/{app,components}/**/*.mock.js'
      ],
      test: {
        src: [
          '<%= yeoman.client %>/{app,components}/**/*.spec.js',
          '<%= yeoman.client %>/{app,components}/**/*.mock.js'
        ]
      }
    },

    // Empties folders to start fresh
    clean: {
      dist: {
        files: [{
          dot: true,
          src: [
            '.tmp',
            '<%= yeoman.dist %>/*',
            '!<%= yeoman.dist %>/.git*',
            '!<%= yeoman.dist %>/.openshift',
            '!<%= yeoman.dist %>/Procfile'
          ]
        }]
      },
      server: {
        files:[{
          dot: true,
          src: [
            '.tmp',
            'report/*',
          ]
        }]
      }
    },

    // Add vendor prefixed styles
    autoprefixer: {
      options: {
        browsers: ['last 1 version']
      },
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/',
          src: '{,*/}*.css',
          dest: '.tmp/'
        }]
      }
    },

    // Debugging with node inspector
    'node-inspector': {
      custom: {
        options: {
          'web-host': 'localhost'
        }
      }
    },

    // Use nodemon to run server in debug mode with an initial breakpoint
    nodemon: {
      debug: {
        script: 'server/app.js',
        options: {
          nodeArgs: ['--debug-brk'],
          env: {
            PORT: process.env.PORT || 9000
          },
          callback: function (nodemon) {
            nodemon.on('log', function (event) {
              console.log(event.colour);
            });

            // opens browser on initial server start
            nodemon.on('config:update', function () {
              setTimeout(function () {
                require('open')('http://localhost:8080/debug?port=5858');
              }, 500);
            });
          }
        }
      }
    },

    // Automatically inject Bower components into the app
    wiredep: {
      target: {
        src: '<%= yeoman.client %>/index.html',
        ignorePath: '<%= yeoman.client %>/',
        exclude: [/bootstrap-sass-official/, /bootstrap.js/, '/json3/', '/es5-shim/']
      }
    },

    // Renames files for browser caching purposes
    rev: {
      dist: {
        files: {
          src: [
            '<%= yeoman.dist %>/public/{,*/}*.js',
            '<%= yeoman.dist %>/public/{,*/}*.css',
            '<%= yeoman.dist %>/public/assets/images/{,*/}*.{png,jpg,jpeg,gif,webp,svg}',
            '<%= yeoman.dist %>/public/assets/fonts/*'
          ]
        }
      }
    },

    // Reads HTML for usemin blocks to enable smart builds that automatically
    // concat, minify and revision files. Creates configurations in memory so
    // additional tasks can operate on them
    useminPrepare: {
      html: ['<%= yeoman.client %>/index.html'],
      options: {
        dest: '<%= yeoman.dist %>/public'
      }
    },

    // Performs rewrites based on rev and the useminPrepare configuration
    usemin: {
      html: ['<%= yeoman.dist %>/public/{,*/}*.html'],
      css: ['<%= yeoman.dist %>/public/{,*/}*.css'],
      js: ['<%= yeoman.dist %>/public/{,*/}*.js'],
      options: {
        assetsDirs: [
          '<%= yeoman.dist %>/public',
          '<%= yeoman.dist %>/public/assets/images'
        ],
        // This is so we update image references in our ng-templates
        patterns: {
          js: [
            [/(assets\/images\/.*?\.(?:gif|jpeg|jpg|png|webp|svg))/gm, 'Update the JS to reference our revved images']
          ]
        }
      }
    },

    // The following *-min tasks produce minified files in the dist folder
    imagemin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.client %>/assets/images',
          src: '{,*/}*.{png,jpg,jpeg,gif}',
          dest: '<%= yeoman.dist %>/public/assets/images'
        }]
      }
    },

    svgmin: {
      dist: {
        files: [{
          expand: true,
          cwd: '<%= yeoman.client %>/assets/images',
          src: '{,*/}*.svg',
          dest: '<%= yeoman.dist %>/public/assets/images'
        }]
      }
    },

    // Allow the use of non-minsafe AngularJS files. Automatically makes it
    // minsafe compatible so Uglify does not destroy the ng references
    ngAnnotate: {
      dist: {
        files: [{
          expand: true,
          cwd: '.tmp/concat',
          src: '*/**.js',
          dest: '.tmp/concat'
        }]
      }
    },

    // Package all the html partials into a single javascript payload
    ngtemplates: {
      options: {
        // This should be the name of your apps angular module
        module: 'edumaterialApp',
        htmlmin: {
          collapseBooleanAttributes: true,
          collapseWhitespace: true,
          removeAttributeQuotes: true,
          removeEmptyAttributes: true,
          removeRedundantAttributes: true,
          removeScriptTypeAttributes: true,
          removeStyleLinkTypeAttributes: true
        },
        usemin: 'app/app.js'
      },
      main: {
        cwd: '<%= yeoman.client %>',
        src: ['{app,components}/**/*.html'],
        dest: '.tmp/templates.js'
      },
      tmp: {
        cwd: '.tmp',
        src: ['{app,components}/**/*.html'],
        dest: '.tmp/tmp-templates.js'
      }
    },

    // Replace Google CDN references
    cdnify: {
      dist: {
        html: ['<%= yeoman.dist %>/public/*.html']
      }
    },

    // Copies remaining files to places other tasks can use
    copy: {
      dist: {
        files: [{
          expand: true,
          dot: true,
          cwd: '<%= yeoman.client %>',
          dest: '<%= yeoman.dist %>/public',
          src: [
            '*.{ico,png,txt}',
            '.htaccess',
            'bower_components/**/*',
            'assets/images/{,*/}*.{webp}',
            'assets/fonts/**/*',
            'index.html'
          ]
        }, {
          expand: true,
          cwd: '.tmp/images',
          dest: '<%= yeoman.dist %>/public/assets/images',
          src: ['generated/*']
        }, {
          expand: true,
          dest: '<%= yeoman.dist %>',
          src: [
            'package.json',
            'server/**/*'
          ]
        }]
      },
      styles: {
        expand: true,
        cwd: '<%= yeoman.client %>',
        dest: '.tmp/',
        src: ['{app,components}/**/*.css']
      }
    },

    buildcontrol: {
      options: {
        dir: 'dist',
        commit: true,
        push: true,
        connectCommits: false,
        message: 'Built %sourceName% from commit %sourceCommit% on branch %sourceBranch%'
      },
      deploy:{
        options: {
            remote: 'production',
            branch: 'master'
          }
      },
      stage:{
        options: {
            remote: 'stage-server',
            branch: 'master'
          }
      },
      heroku: {
        options: {
          remote: 'heroku',
          branch: 'master'
        }
      },
      openshift: {
        options: {
          remote: 'openshift',
          branch: 'master'
        }
      }
    },

    // Run some tasks in parallel to speed up the build process
    concurrent: {
      server: [
      ],
      test: [
      ],
      debug: {
        tasks: [
          'nodemon',
          'node-inspector'
        ],
        options: {
          logConcurrentOutput: true
        }
      },
      dist: [
        'imagemin',
        'svgmin'
      ]
    },

    // Test settings
    karma: {
      unit: {
        configFile: 'karma.conf.js',
        singleRun: true
      }
    },

    mochaTest: {
      options: {
        reporter: 'spec'
      },
      src: ['server/**/*.spec.js']
    },

    protractor: {
      options: {
        configFile: 'protractor.conf.js'
      },
      chrome: {
        options: {
          args: {
            browser: 'chrome'
          }
        }
      }
    },

    env: {
      test: {
        NODE_ENV: 'test'
      },
      prod: {
        NODE_ENV: 'production'
      },
      all: localConfig
    },

    injector: {
      options: {

      },
      // Inject application script files into index.html (doesn't include bower)
      scripts: {
        options: {
          transform: function(filePath) {
            filePath = filePath.replace('/client/', '');
            filePath = filePath.replace('/.tmp/', '');
            return '<script src="' + filePath + '"></script>';
          },
          starttag: '<!-- injector:js -->',
          endtag: '<!-- endinjector -->'
        },
        files: {
          '<%= yeoman.client %>/index.html': [
              ['{.tmp,<%= yeoman.client %>}/{app,components}/**/*.js',
               '!{.tmp,<%= yeoman.client %>}/app/app.js',
               '!{.tmp,<%= yeoman.client %>}/{app,components}/**/*.spec.js',
               '!{.tmp,<%= yeoman.client %>}/{app,components}/**/*.mock.js']
            ]
        }
      },

      // Inject component css into index.html
      css: {
        options: {
          transform: function(filePath) {
            filePath = filePath.replace('/client/', '');
            filePath = filePath.replace('/.tmp/', '');
            return '<link rel="stylesheet" href="' + filePath + '">';
          },
          starttag: '<!-- injector:css -->',
          endtag: '<!-- endinjector -->'
        },
        files: {
          '<%= yeoman.client %>/index.html': [
            '<%= yeoman.client %>/{app,components}/**/*.css'
          ]
        }
      }
    },
  });

  // Used for delaying livereload until after server has restarted
  grunt.registerTask('wait', function () {
    grunt.log.ok('Waiting for server reload...');

    var done = this.async();

    setTimeout(function () {
      grunt.log.writeln('Done waiting!');
      done();
    }, 1500);
  });

  grunt.registerTask('express-keepalive', 'Keep grunt running', function() {
    this.async();
  });

  grunt.registerTask('serve', function (target) {
    if (target === 'dist') {
      return grunt.task.run(['build', 'env:all', 'env:prod', 'express:prod','express-keepalive' ]);
    }

    if (target === 'debug') {
      return grunt.task.run([
        'clean:server',
        'env:all',
        'concurrent:server',
        'injector',
        'wiredep',
        'autoprefixer',
        'concurrent:debug'
      ]);
    }

    grunt.task.run([
      'clean:server',
      
      'env:all',
      'concurrent:server',
      'injector',
      'wiredep',
      'autoprefixer',
      'express:dev',
      'run:metrics',
      'wait',
      // 'open',
      'watch'
    ]);
  });

  grunt.registerTask('server', function () {
    grunt.log.warn('The `server` task has been deprecated. Use `grunt serve` to start a server.');
    grunt.task.run(['serve']);
  });

  grunt.registerTask('test', function(target) {
    if (target === 'server') {
      return grunt.task.run([
        'env:all',
        'env:test',
        'mochaTest'
      ]);
    }

    else if (target === 'client') {
      return grunt.task.run([
        'clean:server',
        'env:all',
        'concurrent:test',
        'injector',
        'autoprefixer',
        'karma'
      ]);
    }

    else if (target === 'e2e') {
      return grunt.task.run([
        'clean:server',
        'env:all',
        'env:test',
        'concurrent:test',
        'injector',
        'wiredep',
        'autoprefixer',
        'express:dev',
        'protractor'
      ]);
    }

    else grunt.task.run([
      'test:server',
      'test:client'
    ]);
  });

  grunt.registerTask('build', [
    'clean:dist',
    'concurrent:dist',
    'injector',
    'wiredep',
    'useminPrepare',
    'autoprefixer',
    'ngtemplates',
    'concat',
    'ngAnnotate',
    'copy:dist',
    'cdnify',
    'cssmin',
    'uglify',
    'changelog:v1',
    'rev',
    'usemin'
  ]);

  grunt.registerTask('default', [
    'newer:jshint',
    'test',
    'build'
  ]);
};
